File: extern/sub_projects/PatchUploader/src/config_parser.cpp
#include "../include/main.hpp"
#include "../include/config_parser.hpp"
#include "../include/worker.hpp"

CConfigParser::CConfigParser() :
	m_bParsed(false)
{
}
CConfigParser::~CConfigParser()
{
}

bool CConfigParser::ParseConfigFile(const std::string& c_stConfigFile)
{	
	try
	{
		LogfA(LOG_FILENAME, "Parsing config file: %s", c_stConfigFile.c_str());

		if (!std::filesystem::exists(c_stConfigFile))
		{
			LogfA(LOG_FILENAME, "Config file does not exist!");
			return false;
		}

		const auto fp = msl::file_ptr(c_stConfigFile, "rb");
		if (!fp)
		{
			LogfA(LOG_FILENAME, "Config file could not open! Error: %d", errno);
			return false;
		}

		const auto buffer = fp.string_read();
		if (buffer.empty())
		{
			LogfA(LOG_FILENAME, "Config file could not read! Error: %d", errno);
			return false;
		}

		doc.Parse<kParseCommentsFlag>(reinterpret_cast<const char*>(buffer.data()));
		if (doc.HasParseError())
		{
			LogfA(LOG_FILENAME, "Config file could not parsed! Error: %s offset: %u", GetParseError_En(doc.GetParseError()), doc.GetErrorOffset());
			return false;
		}
		else if (!doc.IsObject())
		{
			LogfA(LOG_FILENAME, "Config file base is not an object! Type: %u", doc.GetType());
			return false;
		}
		
		if (!doc.HasMember("hosts"))
		{
			LogfA(LOG_FILENAME, "Config 'hosts' key does not exist!");
			return false;
		}
		else if (!doc.HasMember("ftp_hosts"))
		{
			LogfA(LOG_FILENAME, "Config 'ftp_hosts' key does not exist!");
			return false;
		}
		else if (!doc.HasMember("version"))
		{
			LogfA(LOG_FILENAME, "Config 'version' key does not exist!");
			return false;
		}
		else if (!doc.HasMember("sdk_files"))
		{
			LogfA(LOG_FILENAME, "Config 'sdk_files' key does not exist!");
			return false;
		}
		else if (!doc.HasMember("game_files"))
		{
			LogfA(LOG_FILENAME, "Config 'game_files' key does not exist!");
			return false;
		}
		else if (!doc.HasMember("ac_files"))
		{
			LogfA(LOG_FILENAME, "Config 'ac_files' key does not exist!");
			return false;
		}

		if (!doc["hosts"].IsObject())
		{
			LogfA(LOG_FILENAME, "Config 'hosts' key is not object!");
			return false;
		}
		else if (doc["hosts"].ObjectEmpty())
		{
			LogfA(LOG_FILENAME, "Config 'hosts' value is not valid!");
			return false;
		}

		if (!doc["ftp_hosts"].IsObject())
		{
			LogfA(LOG_FILENAME, "Config 'ftp_hosts' key is not object!");
			return false;
		}
		/*
		else if (doc["ftp_hosts"].ObjectEmpty())
		{
			LogfA(LOG_FILENAME, "Config 'ftp_hosts' value is not valid!");
			return false;
		}
		*/

		if (!doc["version"].IsNumber())
		{
			LogfA(LOG_FILENAME, "Config 'version' key is not number!");
			return false;
		}
		else if (!doc["version"].GetInt())
		{
			LogfA(LOG_FILENAME, "Config 'version' value is not valid!");
			return false;
		}

		if (!doc["sdk_files"].IsObject())
		{
			LogfA(LOG_FILENAME, "Config 'sdk_files' key is not object!");
			return false;
		}
		else if (doc["sdk_files"].ObjectEmpty())
		{
			LogfA(LOG_FILENAME, "Config 'sdk_files' value is null!");
			return false;
		}

		if (!doc["game_files"].IsObject())
		{
			LogfA(LOG_FILENAME, "Config 'game_files' key is not object!");
			return false;
		}
		else if (doc["game_files"].ObjectEmpty())
		{
			LogfA(LOG_FILENAME, "Config 'game_files' value is null!");
			return false;
		}

		if (!doc["ac_files"].IsObject())
		{
			LogfA(LOG_FILENAME, "Config 'ac_files' key is not object!");
			return false;
		}
		else if (doc["ac_files"].ObjectEmpty())
		{
			LogfA(LOG_FILENAME, "Config 'ac_files' value is null!");
			return false;
		}

		m_bParsed = true;
		LogfA(LOG_FILENAME, "Config file parsed successfully, processing parsed data...");
		return __ProcessConfigFile(CWorker::Instance().GetParams()->u32PatchVersion);
	}
	catch (const std::exception& e)
	{
		LogfA(LOG_FILENAME, "Exception handled on read_config_file Error: %s", e.what());
		return false;
	}
}

bool CConfigParser::__ProcessConfigFile(uint32_t nVersion)
{
	m_spConfigCtx = std::make_shared<SConfigCtx>();
	if (!m_spConfigCtx)
	{
		LogfA(LOG_FILENAME, "Config context could not be created!");
		return false;
	}

	// S3 config
	if (!doc.HasMember("hosts") || !doc["hosts"].IsObject())
	{
		LogfA(LOG_FILENAME, "hosts key is not valid!");
		return false;
	}

	const auto& hosts = doc["hosts"];
	for (auto it = hosts.MemberBegin(); it != hosts.MemberEnd(); ++it)
	{
		SHostCtx hostCtx;

		if (!it->name.IsString())
		{
			LogfA(LOG_FILENAME, "hosts key is not valid!");
			return false;
		}
		hostCtx.hostname = std::string(it->name.GetString(), it->name.GetStringLength());

		if (!it->value.IsObject())
		{
			LogfA(LOG_FILENAME, "hosts value is not valid!");
			return false;
		}

		const auto& host = it->value;
		if (!host.HasMember("endpoint") || !host["endpoint"].IsString() || !host["endpoint"].GetStringLength())
		{
			LogfA(LOG_FILENAME, "host:endpoint key is not valid!");
			return false;
		}
		hostCtx.endpoint = host["endpoint"].GetString();

		if (!host.HasMember("access_key") || !host["access_key"].IsString() || !host["access_key"].GetStringLength())
		{
			LogfA(LOG_FILENAME, "host:access_key key is not valid!");
			return false;
		}
		hostCtx.access_key = host["access_key"].GetString();

		if (!host.HasMember("secret_key") || !host["secret_key"].IsString() || !host["secret_key"].GetStringLength())
		{
			LogfA(LOG_FILENAME, "host:secret_key key is not valid!");
			return false;
		}
		hostCtx.secret_key = host["secret_key"].GetString();

		m_spConfigCtx->hosts.emplace_back(hostCtx);
	}

	// SFTP config
	if (!doc.HasMember("ftp_hosts") || !doc["ftp_hosts"].IsObject())
	{
		LogfA(LOG_FILENAME, "ftp_hosts key is not valid!");
		return false;
	}

	if (!doc["ftp_hosts"].ObjectEmpty())
	{
		const auto& sftp_hosts = doc["ftp_hosts"];
		for (auto it = sftp_hosts.MemberBegin(); it != sftp_hosts.MemberEnd(); ++it)
		{
			SSFTPHostCtx hostCtx;

			if (!it->name.IsString())
			{
				LogfA(LOG_FILENAME, "ftp_hosts key is not valid!");
				return false;
			}
			hostCtx.hostname = std::string(it->name.GetString(), it->name.GetStringLength());

			if (!it->value.IsObject())
			{
				LogfA(LOG_FILENAME, "ftp_hosts value is not valid!");
				return false;
			}

			const auto& host = it->value;

			if (!host.HasMember("endpoint") || !host["endpoint"].IsString() || !host["endpoint"].GetStringLength())
			{
				LogfA(LOG_FILENAME, "host:endpoint key is not valid!");
				return false;
			}
			hostCtx.endpoint = host["endpoint"].GetString();

			if (!host.HasMember("port") || !host["port"].IsNumber())
			{
				LogfA(LOG_FILENAME, "host:port key is not valid!");
				return false;
			}
			hostCtx.port = host["port"].GetUint();

			if (!host.HasMember("username") || !host["username"].IsString() || !host["username"].GetStringLength())
			{
				LogfA(LOG_FILENAME, "host:username key is not valid!");
				return false;
			}
			hostCtx.username = host["username"].GetString();

			if (!host.HasMember("password") || !host["password"].IsString() || !host["password"].GetStringLength())
			{
				LogfA(LOG_FILENAME, "host:password key is not valid!");
				return false;
			}
			hostCtx.password = host["password"].GetString();

			m_spConfigCtx->sftp_hosts.emplace_back(hostCtx);
		}
	}
	
	// Config version
	m_spConfigCtx->version = doc["version"].GetInt();
	LogfA(LOG_FILENAME, "Version: %d", m_spConfigCtx->version);

	if (m_spConfigCtx->version != 1)
	{
		LogfA(LOG_FILENAME, "Version is not allowed!");
		return false;
	}

	// Root path
	m_spConfigCtx->target_path = fmt::format("{0}\\1.{1}", CWorker::Instance().GetParams()->stRootPath, nVersion);
	LogfA(LOG_FILENAME, "Target path: %s", m_spConfigCtx->target_path.c_str());

	if (!std::filesystem::exists(m_spConfigCtx->target_path))
	{
		LogfA(LOG_FILENAME, "Target path does not exist!");
		return false;
	}

	// Single file
	const auto stUpdateFile = CWorker::Instance().GetParams()->stUpdateFile;
	if (!stUpdateFile.empty())
	{
		const auto stFileNameWithPath = fmt::format("{0}\\{1}", m_spConfigCtx->target_path, stUpdateFile);
		LogfA(LOG_FILENAME, "Single target: %s (%s)", stUpdateFile.c_str(), stFileNameWithPath.c_str());

		if (!std::filesystem::exists(stFileNameWithPath))
		{
			LogfA(LOG_FILENAME, "Target file: %s does not exist!", stFileNameWithPath.c_str());
			return false;
		}

		if (stUpdateFile == "NoMercy_SystemModule_x64.sys")
		{
			auto spFileContainerCtx = std::make_shared<SFileContainerCtx>();
			spFileContainerCtx->id = "ac_files";
			spFileContainerCtx->method = "md5";

			auto spFileCtx = std::make_shared<SFileCtx>();
			spFileCtx->name = stUpdateFile.c_str();
			spFileCtx->local_path = "";
			spFileCtx->attr = NoMercySetup::EFileAttributes::FILE_ATTR_PATH_SYSTEM;
			spFileCtx->optional = false;
			spFileContainerCtx->files.emplace_back(spFileCtx);

			m_spConfigCtx->file_containers.emplace_back(spFileContainerCtx);
			return true;
		}
		else
		{
			LogfA(LOG_FILENAME, "Target file: %s is not a known file!", stFileNameWithPath.c_str());
			return false;
		}
	}

	// Files
	size_t idx = 0;
	for (const auto& objFiles : { doc["sdk_files"].GetObject(), doc["game_files"].GetObject(), doc["ac_files"].GetObject() })
	{
		idx++;

		auto spFileContainerCtx = std::make_shared<SFileContainerCtx>();
		if (idx == 1)
		{
			spFileContainerCtx->id = "sdk_files";
		}
		else if (idx == 2)
		{
			spFileContainerCtx->id = "game_files";
		}
		else if (idx == 3)
		{
			spFileContainerCtx->id = "ac_files";
		}
		else
		{
			LogfA(LOG_FILENAME, "Unknown index: %u", idx);
			return false;
		}

		LogfA(LOG_FILENAME, "Checking object: %d (%s)", idx, spFileContainerCtx->id.c_str());

		size_t file_idx = 0;
		for (auto node = objFiles.MemberBegin(); node != objFiles.MemberEnd(); ++node)
		{
			if (!node->name.IsString())
			{
				LogfA(LOG_FILENAME, "Node: %u key is not an string! Key: %s Type: %u", file_idx, node->name.IsString() ? node->name.GetString() : "<not_string>", node->name.GetType());
				return false;
			}

			const std::string id = node->name.GetString();
			LogfA(LOG_FILENAME, "\t[%u] Node: %s", file_idx, id.c_str());

			if (id == "check_method")
			{
				spFileContainerCtx->method = node->value.GetString();
				LogfA(LOG_FILENAME, "\t\tCheck method: %s", spFileContainerCtx->method.c_str());
				continue;
			}
			else if (id == "files")
			{
				size_t child_idx = 0;
				auto objGameFilesContext = node->value.GetObject();
				for (auto childNode = objGameFilesContext.MemberBegin(); childNode != objGameFilesContext.MemberEnd(); ++childNode)
				{
					child_idx++;

					auto spFileCtx = std::make_shared<SFileCtx>();
					spFileCtx->name = childNode->name.GetString();
					LogfA(LOG_FILENAME, "\t\t[%u] File: %s", child_idx, spFileCtx->name.c_str());

					auto objFileContext = childNode->value.GetObject();
					for (auto fileCtxNode = objFileContext.MemberBegin(); fileCtxNode != objFileContext.MemberEnd(); ++fileCtxNode)
					{
						std::string stFileCtxKey = fileCtxNode->name.GetString();
						if (stFileCtxKey == "local_path")
						{
							spFileCtx->local_path = fileCtxNode->value.GetString();
							LogfA(LOG_FILENAME, "\t\t\tPath: %s", spFileCtx->local_path.c_str());
						}
						else if (stFileCtxKey == "attr")
						{
							for (const auto& pkAttr : fileCtxNode->value.GetArray())
							{
								const auto stAttr = std::string(pkAttr.GetString());
								if (stAttr == "system")
								{
									spFileCtx->attr |= NoMercySetup::EFileAttributes::FILE_ATTR_PATH_SYSTEM;
								}
								else if (stAttr == "game")
								{
									spFileCtx->attr |= NoMercySetup::EFileAttributes::FILE_ATTR_PATH_GAME;
								}
								else if (stAttr == "local")
								{
									spFileCtx->attr |= NoMercySetup::EFileAttributes::FILE_ATTR_PATH_LOCAL;
								}
								else if (stAttr == "hidden")
								{
									spFileCtx->attr |= NoMercySetup::EFileAttributes::FILE_ATTR_HIDDEN;
								}
								else if (stAttr == "crypted_1")
								{
									spFileCtx->attr |= NoMercySetup::EFileAttributes::FILE_ATTR_CRYPTED_1;
								}
								else if (stAttr == "compressed_1")
								{
									spFileCtx->attr |= NoMercySetup::EFileAttributes::FILE_ATTR_COMPRESSED_1;
								}
								else
								{
									LogfA(LOG_FILENAME, "Unknown dependency attribute: %s", stAttr.c_str());
									return false;
								}
								LogfA(LOG_FILENAME, "\t\t\tAttr: %u", spFileCtx->attr);
							}
						}
						else if (stFileCtxKey == "preprocess")
						{
							spFileCtx->preprocess = fileCtxNode->value.GetString();
							LogfA(LOG_FILENAME, "\t\t\tPreprocess: %s", spFileCtx->preprocess.c_str());
						}
						else if (stFileCtxKey == "optional")
						{
							spFileCtx->optional = fileCtxNode->value.GetBool();
							LogfA(LOG_FILENAME, "\t\t\tOptional: %d", spFileCtx->optional ? 1 : 0);
						}
						else if (stFileCtxKey == "skip_pdb")
						{
							spFileCtx->skip_pdb = fileCtxNode->value.GetBool();
							LogfA(LOG_FILENAME, "\t\t\tSkip PDB: %d", spFileCtx->skip_pdb ? 1 : 0);
						}
						else if (stFileCtxKey == "rename")
						{
							spFileCtx->new_name = fileCtxNode->value.GetString();
							LogfA(LOG_FILENAME, "\t\t\tNew name: %s", spFileCtx->new_name.c_str());
						}
						else if (stFileCtxKey == "file_path")
						{
							spFileCtx->local_source_file = fmt::format("{0}\\{1}{2}", m_spConfigCtx->target_path, fileCtxNode->value.GetString(), spFileCtx->name);
							LogfA(LOG_FILENAME, "\t\t\tFile path: %s (%s)", fileCtxNode->value.GetString(), spFileCtx->local_source_file.c_str());
						}
						else
						{
							LogfA(LOG_FILENAME, "Unknown file key: %s", stFileCtxKey.c_str());
							return false;
						}
					}

					spFileContainerCtx->files.emplace_back(spFileCtx);
				}
			}
			else
			{
				LogfA(LOG_FILENAME, "Unknown id: %s", id.c_str());
				return false;
			}

			m_spConfigCtx->file_containers.emplace_back(spFileContainerCtx);
		}
	}

	return true;
};

File: extern/scripts/patch_update/config_parser.py
import os
import jstyleson
import glob
import config_data
import data
from log_helper import log, ERROR_LOG, INFO_LOG

def process_config_file(doc):
	if "hosts" not in doc or not isinstance(doc["hosts"], dict):
		log(ERROR_LOG, "hosts key is not valid!")
		return False

	for hostname, host in doc["hosts"].items():
		hostCtx = {}
		hostCtx["hostname"] = hostname

		if not isinstance(host, dict):
			log(ERROR_LOG, "hosts value is not valid!")
			return False
			
		if "endpoint" in host and isinstance(host["endpoint"], str):
			hostCtx["endpoint"] = host["endpoint"]

		if "access_key" in host and isinstance(host["access_key"], str):
			hostCtx["access_key"] = host["access_key"]

		if "secret_key" in host and isinstance(host["secret_key"], str):
			hostCtx["secret_key"] = host["secret_key"]
   
		if "region" in host and isinstance(host["region"], str):
			hostCtx["region"] = host["region"]

		config_data.hosts.append(hostCtx)
	
	if "ftp_hosts" in doc and isinstance(doc["ftp_hosts"], dict):
		for sftp_host in doc["ftp_hosts"]:
			hostData = doc["ftp_hosts"][sftp_host]		
   
			sftpHostCtx = {}
			sftpHostCtx["name"] = sftp_host
   
			if "hostname" in hostData and isinstance(hostData["hostname"], str):
				sftpHostCtx["hostname"] = hostData["hostname"]
		
			if "endpoint" in hostData and isinstance(hostData["endpoint"], str):
				sftpHostCtx["endpoint"] = hostData["endpoint"]

			if "port" in hostData and isinstance(hostData["port"], int):
				sftpHostCtx["port"] = hostData["port"]

			if "username" in hostData and isinstance(hostData["username"], str):
				sftpHostCtx["username"] = hostData["username"]

			if "password" in hostData and isinstance(hostData["password"], str):
				sftpHostCtx["password"] = hostData["password"]

			if "private_key" in hostData and isinstance(hostData["private_key"], str):
				sftpHostCtx["private_key"] = hostData["private_key"]

			config_data.ftp_hosts.append(sftpHostCtx)

	# Config version
	if "version" in doc and isinstance(doc["version"], int):
		config_data.version = doc["version"]
		log(INFO_LOG, f"Config version: {config_data.version}")

	# Root path
	config_data.target_path = f"{config_data.root_path}{os.path.sep}1.{config_data.patch_version}"
	log(INFO_LOG, f"Target path: {config_data.target_path}")
 
	if not os.path.exists(config_data.target_path):
		log(ERROR_LOG, f"Target path does not exist: {config_data.target_path}")
		return False

	# Single file
	if config_data.update_file:
		update_file_with_path = f"{config_data.target_path}{os.path.sep}{config_data.update_file}"
		log(INFO_LOG, f"Update file: {update_file_with_path}")
	
		if not os.path.exists(update_file_with_path):
			log(ERROR_LOG, f"Update file does not exist: {update_file_with_path}")
			return False

		if config_data.update_file == "NoMercy_SystemModule_x64.sys":
			spFileContainerCtx = data.SFileContainerCtx(id="ac_files", method="md5", files=[])

			spFileCtx = data.SFileCtx(name=config_data.update_file, local_path="", attr=data.EFileAttributes.FILE_ATTR_PATH_SYSTEM, optional=False)
			spFileContainerCtx.files.append(spFileCtx)

			config_data.file_containers.append(spFileContainerCtx)
			return True
		else:
			log(ERROR_LOG, f"Update file is not supported: {config_data.update_file}")
			return False

	# Multiple files
	if "sdk_files" not in doc or not isinstance(doc["sdk_files"], dict):
		log(ERROR_LOG, "sdk_files key is not valid!")
		return False
	elif "game_files" not in doc or not isinstance(doc["game_files"], dict):
		log(ERROR_LOG, "game_files key is not valid!")
		return False
	elif "ac_files" not in doc or not isinstance(doc["ac_files"], dict):
		log(ERROR_LOG, "ac_files key is not valid!")
		return False
 
	idx = 0
	for objFiles in [doc["sdk_files"], doc["game_files"], doc["ac_files"]]:
		idx += 1
  
		containerId = ""
		if idx == 1:
			containerId = "sdk_files"
		elif idx == 2:
			containerId = "game_files"
		elif idx == 3:
			containerId = "ac_files"
		else:
			log(ERROR_LOG, f"Invalid container id: {idx}")
			return False

		# log(INFO_LOG, f"Checking object: {idx} ({containerId}) > {objFiles}")

		spFileContainerCtx = data.SFileContainerCtx(containerId, method=objFiles["check_method"], files=[])
  
		file_idx = 0
		for fileName in objFiles["files"]:
			file_idx += 1
   
			fileData = objFiles["files"][fileName]
			log(INFO_LOG, f"Checking file: {file_idx} > {fileName} > {fileData}")
   
			fileAttr = data.EFileAttributes.FILE_ATTR_NONE.value
			localSourceFile = ""
      
			if "attr" in fileData:
				for attr in fileData["attr"]:
					if attr == "system":
						fileAttr |= data.EFileAttributes.FILE_ATTR_PATH_SYSTEM.value
					elif attr == "game":
						fileAttr |= data.EFileAttributes.FILE_ATTR_PATH_GAME.value
					elif attr == "local":
						fileAttr |= data.EFileAttributes.FILE_ATTR_PATH_LOCAL.value
					elif attr == "hidden":
						fileAttr |= data.EFileAttributes.FILE_ATTR_HIDDEN.value
					elif attr == "crypted_1":
						fileAttr |= data.EFileAttributes.FILE_ATTR_CRYPTED_1.value
					elif attr == "compressed_1":
						fileAttr |= data.EFileAttributes.FILE_ATTR_COMPRESSED_1.value
					else:
						log(ERROR_LOG, f"Invalid file attribute: {attr}")
						return False
				if not fileAttr:
					log(ERROR_LOG, f"Invalid file attribute: {fileData['attr']}")
					return False
			if "file_path" in fileData:
				log(INFO_LOG, f"Specific file path found: {fileData['file_path']} for {fileName}")
    
				localSourceFile = f"{config_data.target_path}{os.path.sep}{fileData['file_path']}{fileName}"
				if not "*" in localSourceFile:
					if not os.path.exists(localSourceFile):
						log(ERROR_LOG, f"Local source file does not exist: {localSourceFile}")
						return False
				else:
					matchedFiles = glob.glob(localSourceFile)
					if not matchedFiles:
						log(ERROR_LOG, f"Local source file does not exist: {localSourceFile}")
						return False
   
			fileCtx = data.SFileCtx(
       			name = fileName,
          		local_path = fileData["local_path"],
            	attr = fileAttr,
				preprocess = fileData["preprocess"] if "preprocess" in fileData else "",
             	optional = fileData["optional"],
				skip_pdb = fileData["skip_pdb"] if "skip_pdb" in fileData else False,
				new_name = fileData["rename"] if "rename" in fileData else "",
				local_source_file = localSourceFile
            )
   
			# log(INFO_LOG, f"File context: {fileName} >> {localSourceFile}")
			spFileContainerCtx.files.append(fileCtx)
		
		config_data.file_containers.append(spFileContainerCtx)
  
	# print(f"File containers: {config_data.file_containers}")
	log(INFO_LOG, f"Processed file container count: {len(config_data.file_containers)}")
	return True
				

def parse_config_file():
	log(INFO_LOG, f"Parsing config file: {config_data.config_file_name}")

	if not os.path.exists(config_data.config_file_name):
		log(ERROR_LOG, "Config file does not exist!")
		return False

	with open(config_data.config_file_name, 'r') as fp:
		buffer = fp.read()
		if not buffer:
			log(ERROR_LOG, "Config file could not read!")
			return False
		
		doc = jstyleson.loads(buffer)
	
		if "hosts" not in doc:
			log(ERROR_LOG, "Config 'hosts' key does not exist!")
			return False
		elif "ftp_hosts" not in doc:
			log(ERROR_LOG, "Config 'ftp_hosts' key does not exist!")
			return False
		elif "version" not in doc:
			log(ERROR_LOG, "Config 'version' key does not exist!")
			return False
		elif "sdk_files" not in doc:
			log(ERROR_LOG, "Config 'sdk_files' key does not exist!")
			return False
		elif "game_files" not in doc:
			log(ERROR_LOG, "Config 'game_files' key does not exist!")
			return False
		elif "ac_files" not in doc:
			log(ERROR_LOG, "Config 'ac_files' key does not exist!")
			return False

		if not isinstance(doc["hosts"], dict):
			log(ERROR_LOG, "Config 'hosts' key is not object!")
			return False
		elif not doc["hosts"]:
			log(ERROR_LOG, "Config 'hosts' value is not valid!")
			return False

		if not isinstance(doc["ftp_hosts"], dict):
			log(ERROR_LOG, "Config 'ftp_hosts' key is not object!")
			return False
		# elif not doc["ftp_hosts"]:
		# 	log(ERROR_LOG, "Config 'ftp_hosts' value is not valid!")
		# 	return False

		if not isinstance(doc["version"], int):
			log(ERROR_LOG, "Config 'version' key is not number!")
			return False
		elif not doc["version"]:
			log(ERROR_LOG, "Config 'version' value is not valid!")
			return False

		if not isinstance(doc["sdk_files"], dict):
			log(ERROR_LOG, "Config 'sdk_files' key is not object!")
			return False
		elif not doc["sdk_files"]:
			log(ERROR_LOG, "Config 'sdk_files' value is null!")
			return False

		if not isinstance(doc["game_files"], dict):
			log(ERROR_LOG, "Config 'game_files' key is not object!")
			return False
		elif not doc["game_files"]:
			log(ERROR_LOG, "Config 'game_files' value is null!")
			return False

		if not isinstance(doc["ac_files"], dict):
			log(ERROR_LOG, "Config 'ac_files' key is not object!")
			return False
		elif not doc["ac_files"]:
			log(ERROR_LOG, "Config 'ac_files' value is null!")
			return False

		log(INFO_LOG, "Config file parsed successfully, processing parsed data...")
		return process_config_file(doc)
